service/matching/matchingEngine.go

```diff
@@ -961,14 +1018,44 @@ func (e *matchingEngineImpl) ApplyTaskQueueUserDataReplicationEvent(
 		return nil, err
 	}
 	updateOptions := UserDataUpdateOptions{
-		Replicate: false,
 		// Avoid setting a limit to allow the replication event to always be applied
 		TaskQueueLimitPerBuildId: 0,
 	}
-	err = tqMgr.UpdateUserData(ctx, updateOptions, func(current *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, error) {
+	err = tqMgr.UpdateUserData(ctx, updateOptions, func(current *persistencespb.TaskQueueUserData) (*persistencespb.TaskQueueUserData, bool, error) {
 		mergedUserData := *current
-		mergedUserData.VersioningData = MergeVersioningData(current.GetVersioningData(), req.GetUserData().GetVersioningData())
-		return &mergedUserData, nil
+		_, buildIdsRemoved := GetBuildIdDeltas(current.GetVersioningData(), req.GetUserData().GetVersioningData())
+		var buildIdsToRevive []string
+		for _, buildId := range buildIdsRemoved {
+			exists, err := worker_versioning.WorkflowsExistForBuildId(ctx, e.visibilityManager, ns, req.TaskQueue, buildId)
```

Comment:
Should this have any special rate limiting applied?
Looks like visibility manager already enforces rate limits.